### DISCUSSION TITLE
how to add type hint for pyodide.JsProxy object in a external Python file

### DISCUSSION THREAD

**Discussion by grimmerk** [pyodide/pyodide#1833] [Q&A] [answered]
I put my Python code in external Python files and use Visual Studio Code as my main IDE. It would be useful to use type hints and VSCode can help me to maintain the code. 

`JsProxy` is not explicitly imported in this class so that I can not use it for typing. Is there any suggestion? 

BTW, I have some questions about `destroy` and to avoid creating too many issues/discussions so just ask here together. 

1. Why `JsProxy` does not need `destroy()` but `PyProxy` does need?

2. I guess `JsProxy` is in Python wasm memory and `PyProxy` is in JS memory. Originally I thought `PyProxy`' `destroy()` is something like to remove any its reference so garbage collection can collect it. But 
_**If the browser supports FinalizationRegistry then a `PyProxy` that is not part of a Javascript/Python reference cycle will eventually be collected**_
seems to indicate it is not like this. As far as I know, there is no force and immediately destroying method in JS. I do not quite understand this part. I have similar doubts about `PyBuffer`. 

3. **_Furthermore, the Javascript garbage collector does not have any information about whether Python is experiencing memory pressure. So it’s best to aim to avoid leaks._** 
does it mean that PyProxy may access some Python objects which are garage collected by Python? Or it means that `PyProxy` will prevent Python objects been garbage collected and destroy() will minus one reference count of the reference Python object?

4. I encounter a `Comlink` unclonable issue when returning PyProxy data to main threads, finally, I use `Comlink.proxy(somePyProxyData)` to work around it (but `getBuffer()` is not working and only `toJs()` works).  This data includes `ndarray` and `bytes` and not cloneable. I notice  there is a method `pyodide.registerComlink`
and is it about this issue? Is any detailed introuduction to this method?

5. https://pyodide.org/en/stable/usage/type-conversions.html#round-trip-conversions roughly I understand but how do I exactly do some kind of `translation back unwraps the proxy`?

**Comment by hoodmane** [comment, marked as answer]
Thanks for your questions @grimmer0125. Here are some answers:

> `JsProxy` is not explicitly imported in this class so that I can not use it for typing. Is there any suggestion?

I don't understand your question, can you elaborate?

> 1. Why `JsProxy` does not need `destroy()` but `PyProxy` does need?

`JsProxy` doesn't need `destroy` because the Python garbage collector destroys things immediately when their reference count hits zero. So the correct way to destroy the `JsProxy` is to delete all references to it, the destructor will be immediately invoked. `PyProxy` needs the `destroy` method because even if all references to the `PyProxy` are removed, the `PyProxy` can hang around for a very long time because of the way the browser garbage collector works.

> 2.  I guess JsProxy is in Python wasm memory and PyProxy is in JS memory. 

Yes.

> I thought PyProxy' destroy() is something like to remove its reference

Yes. We attach a finalization callback that will call the `destroy` method when the `PyProxy` is collected. If `destroy` is called, we remove that finalization callback from the `PyProxy` because it is no longer needed. Perhaps the wording could be improved here.

> 3. it means that PyProxy will prevent Python objects been garbage collected and destroy() will minus one reference count of the reference Python object?

Yes.

> 4. I encounter a `Comlink` unclonable issue when returning `PyProxy` data to main threads, finally, I use `Comlink.proxy(somePyProxyData)` to work around it (but `getBuffer()` is not working and only `toJs()` works). This data includes `ndarray` and `bytes` and not cloneable.

This is complicated, please open an issue with the same question and we can discuss. Eventually we want to have a documented solution for this as part of Pyodide but we aren't there yet. I think it will require actual software design not just docs though. Your input is appreciated.

> I notice there is a method `pyodide.registerComlink`
and is it about this issue?

No

> Is any detailed introuduction to this method?

No.

> 5. how do I exactly do some kind of "translation back unwraps the proxy"?

A `PyProxy` is a Javascript object which wraps a Python object. If you translate the `PyProxy` into Python, it will turn into the Python object that it wraps.

Conversely, a `JsProxy` is a Python object that wraps a Javascript object. If you translate it into Javascript, it will turn into the Javascript object that it wraps.

**Reply by grimmerk** [reply]
@hoodmane  thank you for the answers !!

The use case in my first question is 

Browser TypeScript file:
```ts
// app.ts
let PyodideTestClass = null
async function step1_initPyodide(){
  const pythonCode = await (await fetch(baseURL+'python/test_class.py')).text();
  await self.pyodide.loadPackagesFromImports(pythonCode);
  await self.pyodide.runPythonAsync(pythonCode); 
  // Get constructor function
  PyodideTestClass = self.pyodide.globals.get('TestClass')
}

async function step2_newTestClass(buffer:ArrayBuffer, decoder: any) {
  // decoder is a JS object and becomes JsProxy object in Python
  const TestClassInstancePyProxy= PyodideTestClass(buffer, decoder);
}
```
external Python file
```py
# test_class.py
class TestClass:
    def __init__(self, buffer: Any = None, decoder: Any = None):
        pass
```
I want to type `buffer:pyodide.JsProxy` and decoder as well so that Python intelligence of VSCode can identify them and improve maintenance.  The question is I do not explicitly import pyodide in this file so I can not type annotate them.

**Reply by hoodmane** [reply]
Well the file `pyodide.d.ts` has type definitions. I think you can do `import PyProxy from 'pyodide_build_path/pyodide.d'` and get the behavior you are looking for.

**Reply by grimmerk** [reply]
Thank you. It's good news that we can use `PyProxy` on a TypeScript side.

On the Python side, how do we annotate  `buffer` is `JsProxy` in  `def __init__(self, buffer: Any = None, decoder: Any = None)`?

**Reply by hoodmane** [reply]
Add `from pyodide import JsProxy` to the top of the file, then annotate with `JsProxy`. 

We don't currently have annotations for subtypes of `JsProxy`. For instance, a buffer has extra methods `buf.assign_to` and `buf.assign` to copy the buffer data into and out of Python's linear memory. It might be nice to have e.g., a `JsBuffer` type that knows about these methods. If you need this, please open an issue. I think the design of anything that works well for this will be a bit complicated.

**Reply by grimmerk** [reply]
I tried to add `from pyodide import JsProxy` and annotate with JsProxy. The script is executeded well. However
1. My VSCode still show the parameter type is Any when the mouse hovers it. 
2. A warning: `Import "pyodide" could not be resolved` 
3. No auto-completion from IDE intelligence, such as showing `to_py()` when typing `buffer.to_py()`. 

I think all are because `pyodide` is not existing local `PYTHONPATH` when editing this Python script. It dynamically exists in Python WASM in runtime.

**Reply by hoodmane** [reply]
Indeed. It is possible to `import pyodide` in native Python builds. In this environment, the ordinary code for `JsProxy` is missing (since it cannot be build outside of emscripten) but we provide type stubs so that type mypy etc should be able to type check code anyways.

We should release the pure Python code as a pypi package. If we did this, then you could do `pip install pyodide` and then this would work. Probably we should warn that it's mostly only useful for type checkers and testing.

In the meantime, I think you can create a `setup.py` file in `src/py` with just:
```py
from setuptools import setup
if __name__ == "__main__":
    setup()
```
and then do:
```py
pip install -e src/py
```
Assuming you do this in the same virtual environment as VSCode is using, I think that should fix these problems.

**Reply by hoodmane** [reply]
Opened #1839.
